src/main/scala/com/microsoft/hyperspace/index/rules/JoinIndexRule.scala

```diff
@@ -68,18 +72,54 @@ object JoinIndexRule
                     right =
                       RuleUtils.transformPlanToUseIndex(spark, rIndex, r, useBucketSpec = true))
 
+              def getShuffleCnt(sparkPlan: SparkPlan): Long = {
+                sparkPlan.collect { case _: ShuffleExchangeExec => true }.length
+              }
+
+              val resultPlan =
+                if (!HyperspaceConf.hybridScanEnabled(spark) || !HyperspaceConf
+                      .hybridScanShuffleCheckEnabled(spark)) {
+                  updatedPlan
+                } else {
+                  val shuffleCntPair = spark.sessionState
```

Comment:
is it possible that this will be recursively called since this is called inside a rule? (I am not sure if this is a good pattern or not to be honest).

Reply:
Good point. how about just using `JoinSelection` & and just check the bucket number of candidate index pair if it's SortMergeJoin?